File: src/0325-Maximum-Size-Subarray-Sum-Equals-k/0325.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution 
{
public:
    int maxSubArrayLen(vector<int>& nums, int k) 
    {
        int sum = 0, res = 0;
        unordered_map<int, int> m;
        for (int i = 0; i < nums.size(); ++i) 
        {
            sum += nums[i];
            if (sum == k) res = i + 1;
            else if (m.count(sum - k)) res = max(res, i - m[sum - k]);
            if (!m.count(sum)) m[sum] = i;
        }
        return res;
    }
};

File: src/0325-Maximum-Size-Subarray-Sum-Equals-k/0325.py
class Solution:
    def maxSubArrayLen(self, nums, k):
        """
        :type nums: list
        :type k: int
        :rtype: int
        """
        pre_sum, res = 0, 0
        dic = dict()

        for i in range(len(nums)):
            pre_sum += nums[i]
            if pre_sum == k:
                res = i + 1
            elif pre_sum - k in dic:
                res = max(res, i - dic[pre_sum - k])
            if pre_sum not in dic:
                dic[pre_sum] = i

        return res

if __name__ == '__main__':
    nums = [1, -1, 5, -2, 3]
    k = 3
    print(Solution().maxSubArrayLen(nums, k))